File: core/client/game/src/commonGame/control/KeyboardControl.cs
using System;
using ShineEngine;
using UnityEngine;

/// <summary>
/// 键盘控制
/// </summary>
[Hotfix]
public class KeyboardControl:SBaseEventRegister<KeyCode>
{
	public void init()
	{
		SKeyboardControl.keyFunc+=onSKey;
	}

	public bool isKeyDown(KeyCode code)
	{
		return SKeyboardControl.isKeyDown(code);
	}

	private void onSKey(KeyCode code,bool isDown)
	{
		onKey(code,isDown);

		dispatch(isDown ? KeyEventType.KeyDown : KeyEventType.KeyUp,code);
	}

	/** 键盘操作 */
	protected virtual void onKey(KeyCode code,bool isDown)
	{
		switch(code)
		{
			case KeyCode.Escape:
			{
				if(!isDown)
				{
					escDown();
				}
			}
				break;
		}
	}

	/** esc退出方法 */
	protected virtual bool escDown()
	{
        if (GameC.nativeUI == null)
            return false;
        Reporter reporter = GameC.nativeUI.getReporter();
        
		if(CommonSetting.useReporter && reporter!=null && reporter.show)
		{
			reporter.hideReporter();
			return true;
		}
        GMCommandUI gMCommandUI = GameC.nativeUI.getGMCommandUI();
        if (CommonSetting.useDebug && gMCommandUI != null && gMCommandUI.isShow())
		{
			gMCommandUI.hide();
			return true;
		}
        if (CommonSetting.useReporter && reporter!=null && !reporter.show)
		{
			reporter.showReporter();
			return true;
		}
        //其他

        return false;
	}
}

File: core/clientLaya/game/src/commonGame/control/KeyboardControl.ts
namespace Shine
{
	export class KeyboardControl extends SBaseEventRegister<number>
	{
		constructor()
        {
            super();
        }

	    public init():void
		{
			SKeyboardControl.keyFunc=Func.create(this,this.onSKey);
		}
	
		private onSKey(code:number,isDown:boolean):void
		{
			this.onKey(code,isDown);
	
			this.dispatch(isDown ? KeyEventType.KeyDown : KeyEventType.KeyUp,code);
		}
	
		/** 键盘操作 */
		protected onKey(code:number,isDown:boolean):void
		{
			switch(code)
			{
				case Laya.Keyboard.ESCAPE:
				{
					if(!isDown)
					{
						this.escDown();
					}
				}
					break;
				case Laya.Keyboard.NUMPAD_ADD:
				{
					if(!isDown)
					{
						this.escDown();
					}
				}
					break;
			}
		}
	
		/** esc退出方法 */
		protected escDown():boolean
		{
			// if(CommonSetting.useReporter && GameC.debug.getReporter().show)
			// {
			// 	GameC.debug.getReporter().hideReporter();
			// 	return true;
			// }
	
			// if(CommonSetting.useDebug && GameC.debug.getGMCommandUI().isShow())
			// {
			// 	GameC.debug.getGMCommandUI().hide();
			// 	return true;
			// }
	
			// if(CommonSetting.useReporter && !GameC.debug.getReporter().show)
			// {
			// 	GameC.debug.getReporter().showReporter();
			// 	return true;
			// }
	
			//其他
	
			return false;
		}
	}
}